src/lakefs_spec/transaction.py

```diff
@@ -29,112 +27,88 @@
 if TYPE_CHECKING:  # pragma: no cover
     from lakefs_spec import LakeFSFileSystem
 
-    VersioningOpTuple = tuple[Callable[[Client], None], str | "Placeholder" | None]
 
-
-@dataclass
-class Placeholder(Generic[T], wrapt.ObjectProxy):
-    """A generic placeholder for a value computed by the lakeFS server in a versioning operation during a transaction."""
-
-    def __init__(self, wrapped: T | None = None):
-        super().__init__(wrapped)
-
-    @property
-    def available(self) -> bool:
-        """Whether the wrapped value is available, i.e. already computed."""
-        return self.__wrapped__ is not None
-
-    @property
-    def value(self):
-        if self.__wrapped__ is None:
-            raise RuntimeError("placeholder unfilled")
-        return self.__wrapped__
-
-    @value.setter
-    def value(self, val: T) -> None:
-        """Fill in the placeholder. Not meant to be called directly except in the completion of the transaction."""
-        self.__wrapped__ = val
+def ensurebranch(b: str | Branch, repository: str, client: Client) -> Branch:
+    if isinstance(b, str):
+        return Branch(repository, b, client=client)
+    return b
 
 
 class LakeFSTransaction(Transaction):
     """
     A lakeFS transaction model capable of versioning operations in between file uploads.
 
+    Creates an ephemeral branch, conducts all uploads and operations on that branch,
+    and optionally merges it back into the source branch on success.
+
     Parameters
     ----------
     fs: LakeFSFileSystem
         The lakeFS file system associated with the transaction.
+    repository: str | Repository
+        The repository in which to conduct the transaction.
+    base_branch: str | Branch
+        The branch on which the resulting files should end up.
+    automerge: bool
+        Automatically merge the ephemeral branch into the base branch after successful
+        transaction completion.
+    delete: bool
+        Delete the ephemeral branch after the transaction.
     """
 
-    def __init__(self, fs: "LakeFSFileSystem"):
+    def __init__(
+        self,
+        fs: "LakeFSFileSystem",
+        repository: str | Repository,
+        base_branch: str | Branch = "main",
+        automerge: bool = True,
+        delete: bool = True,
```

Comment:
I wonder if the default should be `False` (or the behavior modified to only delete on successful transactions). If something goes wrong in the transaction, the user might appreciate the branch still being around for troubleshooting (and it might be hard to reproduce the issue when re-running the code after setting it to `False`).

Reply:
Resolved in [caa9a0a](https://github.com/aai-institute/lakefs-spec/pull/258/commits/caa9a0a9c1da8b8065f2bdd480a41f3d42b5b99e) by making it a three-option literal as discussed in DM.